### ISSUE TITLE
Implement cross platform EventPipeProfiler diagnoser

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/BenchmarkDotNet#1315]
`EtwProfiler` is very popular, but works only on Windows. .NET Core 3.0 shipped with some new cool diagnostics features and tools. We can use the new diagnostics tooling to allow for profiling benchmarks on Linux and macOS.

We have two options:

* use [Microsoft.Diagnostics.NETCore.Client](https://github.com/dotnet/diagnostics/pull/617) library
* use [dotnet trace](https://github.com/dotnet/diagnostics/tree/master/src/Tools/dotnet-trace) global tool

Microsoft.Diagnostics.NETCore.Client has not been published to nuget.org yet, so the person who is willing to work on this could start with `dotnet trace`.

The new diagnoser should produce a speedscope file by default.

/cc @noahfalk @WojciechNagorski

**Comment by WojciechNagorski** [comment]
I would like to do it. I thought about dotnet trace, as you know 😉

**Comment by adamsitnik** [comment]
> I would like to do it.

Awesome! I've assigned you :)

**Comment by WojciechNagorski** [comment]
I have been doing the `EventPipeProfiler` and I've found a problem with `dotnet trace`. There is an issue https://github.com/dotnet/diagnostics/issues/645

When I run `dotnet trace` in the same console without output redirection I get an output file. 
So currently this issue is blocked.

**Comment by sywhang** [comment]
@WojciechNagorski `Microsoft.Diagnostics.NETCore.Client` is now available as a nuget package here: https://www.nuget.org/packages/Microsoft.Diagnostics.NETCore.Client/0.1.0-preview.19606.2

Thought I'd let you know since I know you were blocked by this :)

**Comment by AndreyAkinshin** [comment]
Unfortunately, we can't reference prerelease NuGet packages in the release versions of BenchmarkDotNet. @sywhang any estimates for the release version of `Microsoft.Diagnostics.NETCore.Client`?

**Comment by WojciechNagorski** [comment]
@sywhang Thanks! I'm going to use it to develop `EventPipeProfiler` but, as @AndreyAkinshin wrote, we should not use prerelease NuGet because of https://github.com/NuGet/Home/issues/2944.

**Comment by WojciechNagorski** [comment]
@AndreyAkinshin Did you noticed that I created new NuGet package `BenchmarkDotNet.Diagnostics.NETCore` with .NETCoreApp 2.1 target because of `Microsoft.Diagnostics.NETCore.Client` support only .NETCoreApp 2.1 target. 

Couldn't we publish a new `BenchmarkDotNet.Diagnostics.NETCore` NuGet with reference to the pre-release version of `Microsoft.Diagnostics.NETCore.Client`? This new package could also have a pre-release version, e.g. 0.12.1-prereview1.

**Comment by AndreyAkinshin** [comment]
@WojciechNagorski to be honest, I don't like this idea. We already had such an experience of separate prerelease versions for some specific packages. It wasn't a pleasant experience because of the two following reasons:
1. Our infrastructure and release workflow become more complicated because we have to manage two different versions of our packages (release and prerelease).
2. Users are getting confused because they think that `0.12.1-preview*` is a version before `0.12.1`. Once BenchmarkDotNet `0.12.1` is released, nobody expects that they need to install `0.12.1-preview*` to enable additional features in `0.12.1`.

**Comment by WojciechNagorski** [comment]
@AndreyAkinshin You're right. I have to agree with your points.

**Comment by sywhang** [comment]
@AndreyAkinshin 

> any estimates for the release version of Microsoft.Diagnostics.NETCore.Client?

The preview tag is mainly there because I didn't want to put it out officially until enough customers have tried it and found it useful and validated it. Since this PR is one of the first few use cases of it, I'm hoping to remove the preview tag and push out the release version pretty soon (before end of the year).

**Comment by adamsitnik** [comment]
> before end of the year

@sywhang this is great news! The most important thing for us it to have a NuGet package that is not marked as pre-release to avoid https://github.com/NuGet/Home/issues/2944#issuecomment-563416633. It can be called "beta" or "preview" or anything like this (if this helps on your side in any way)

**Comment by WojciechNagorski** [comment]
@sywhang @adamsitnik So I'm going to finish this task this or next week.

**Comment by sywhang** [comment]
@adamsitnik @WojciechNagorski 
Release version of Microsoft.Diagnostics.NETCore.Client has been just pushed to NuGet:
https://www.nuget.org/packages/Microsoft.Diagnostics.NETCore.Client/0.2.61701 

Let me know if there's anything I can help you with to get you unblocked!

**Comment by adamsitnik** [comment]
@sywhang this is great news, thank you!

**Comment by WojciechNagorski** [comment]
@sywhang @adamsitnik this task is not blocked anymore. I've just pushed new version. I have only one blocker it is my work. Currently benchmarkdotnet is only my hobby. I wish I could work on full time here 😉

**Comment by MrGuardian** [comment]
Any news on this issue? I'm very interested in BenchmarksDotNet working on MacOS and seeing cache misses.

**Comment by WojciechNagorski** [comment]
@MrGuardian I've done it here #1321. I'm waiting for @adamsitnik's review.

**Comment by MrGuardian** [comment]
@adamsitnik I wonder when this feature is finally gonna be merged? With .netcore being essentially cross-platform we need a tool working on non-Windows machines to measure performance.

**Comment by MrGuardian** [comment]
Thank you @WojciechNagorski and @adamsitnik for merging #1321 and your incredible work!

Yesterday I downloaded the latest master branch, added BenchmarkDotNet and BenchmarkDotNet.Annotations projects as dependencies to my demo project (Md5VsSha256 tests from you documentation basically) on MacOS.

I marked the whole class with the following attribute `[HardwareCounters(HardwareCounter.CacheMisses)]` expecting BenchmarkDotNet to produce the information about cache misses on my machine. But in return, the benchmarks output just the baseline information and don't have cache misses statistics.

It is worthy to note that at the start of the benchmarking session the logs prints this error:

```
// Validating benchmarks:
Unable to resolve IHardwareCountersDiagnoser diagnoser using dynamic assembly loading. 
            Please make sure that it's supported on macOS 10.15.3 (19D76) [Darwin 19.3.0]
```

so maybe that's why I can't gather advanced CPU metrics. I thought this PR would solve this. The program also crashes after producing the output for the tests.

Here is basically the whole code for my project: https://gist.github.com/MrGuardian/56bb63a0ea79867d4e5fdd67894fd876

Here is the machine I use based on the report of BenchmarkDotNet:
```
BenchmarkDotNet=v0.12.0.20200306-develop, OS=macOS 10.15.3 (19D76) [Darwin 19.3.0]
Intel Core i7-8850H CPU 2.60GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.100
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), X64 RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), X64 RyuJIT
```

Basically Macbook Pro 15 2018.

**Comment by adamsitnik** [comment]
> I thought this PR would solve this. 

the goal of this PR was to add a simple cross-platform CPU profiler. It's internally using `EventPipe` mechanism which is something like `ETW` but implemented inside `CoreCLR`. It can give us CPU samples and .NET EventSource events, but it does not expose any hardware counters information.

As of today, the only way to get hardware counters information for .NET Core on macOS is to use Intel VTune profiler. It's really awesome tool and it's now available for free.

**Comment by adamsitnik** [comment]
Fixed in #1321 by @WojciechNagorski
